### PR TITLE
Migrate DeploymentCreateProcessor to generalised distribution logic

### LINKED ISSUE #11661: Switch `Deployment:Create` processor to new record distribution behavior
- Switching should be backward compatible
- We don't delete the existing processors for `Deployment` commands
- We don't delete the existing appliers for `Deployment` events
- Switching happens by changing the processor's logic
- ~~We also need to switch the logic in the `DeploymentRedistributor`~~
- `DeploymentCreateProcessor` should be idempotent

Depends on
- #11660

### PR DESCRIPTION
**Pull request by remcowesterhoud**
## Description

Brace yourselves, this is large one. I did my best to split it all in small commits. I recommend reviewing it by commit as well.

This PR migrates the `DeploymentCreateProcessor` from the old distribution logic to the generalised distribution logic. This involves several parts:

- We need to be able to identify if a Deployment CREATE command came from another partition, or if it was send by a client
- We need to migrate the tests related to deployment distributions as the records are different from before
- Before resources were all stored using the DeploymentDistribution DISTRIBUTED applier. This will no longer be used and will now be done using Process/DecisionRequirements/Decision CREATED events instead.
- We need to start using the `CommandDistributionBehavior` for the distributions

## Related issues

closes #11661 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [x] There are unit/integration tests that verify all acceptance criterias of the issue
* [x] New tests are written to ensure backwards compatibility with further versions
* [x] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Other teams:
If the change impacts another team an issue has been created for this team, explaining what they need to do to support this change.
- [ ] [Operate](https://github.com/camunda/operate/issues)
- [ ] [Tasklist](https://github.com/camunda/tasklist/issues)
- [ ] [Web Modeler](https://github.com/camunda/web-modeler/issues)
- [ ] [Desktop Modeler](https://github.com/camunda/camunda-modeler/issues)
- [ ] [Optimize](https://github.com/camunda/camunda-optimize/issues)

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

**Comment by remcowesterhoud** [comment]
@korthout This thing is huge so please take your time to review this. I'm happy to go through it together if it helps.

**Comment by remcowesterhoud** [comment]
I've unrequested you review, seems there is still a bunch of failing tests 😞

**Comment by korthout** [comment]
@remcowesterhoud Do you want to rebase this, or shall I?

**Comment by remcowesterhoud** [comment]
I'll do it, I think most of the changes in this PR are mine anyway.

**Comment by korthout** [comment]
@remcowesterhoud One thing we could do to reduce the size of this pull request, is to extract the following into a or several separate pull request(s):
- add the ability to records to see if they are distributed or not (6fd299e)
- acknowledge command in command distribution behavior (58e4a13)
- allow easy filtering of records of the type CommandDistribution (c045477)
- add `partitionId` to empty command distribution records (751881b)
- add command distribution value type to TypedEventRegistry (3e4dfe8)
- distribute a copy of command value (854e30f)
- return a unified record value (51c38e1)
- return null when ValueType is NULL_VAL (e2dd4f80)
- prevent limiting of CommandDistribution ACKNOWLEDGE commands (3b1db311)

All these are not specifically related to migrating deployment distribution, but rather on making generalized record distribution usable.

These could enable us to start work on distributing signals, without having to wait for deployment distribution to be migrated.

**Comment by remcowesterhoud** [comment]
Yes I agree, I was also looking into why this was so huge 😅

**Comment by korthout** [comment]
Let me know if you need a hand extracting these commits

**Comment by korthout** [comment]
>@korthout changed the base branch from main to korthout-11456-command-distribution-improvements [now](https://github.com/camunda/zeebe/pull/11962#event-9445333811)

I did this so we only see the commits this pull request adds on top of #12970.

I'm curious whether the base branch automatically updates to `main` when we merge `korthout-11456-command-distribution-improvements` into `main`.

**Comment by remcowesterhoud** [comment]
> Base automatically changed from korthout-11456-command-distribution-improvements  to main  

🎉

**Review by korthout — CHANGES_REQUESTED** [review]
❌ There are some test failures. Do you want to look into them together?

```
main_test.go:267: failed to reformat response JSON: invalid character 'E' looking for beginning of value
        Erroneous JSON: Error: rpc error: code = InvalidArgument desc = Command 'EVALUATE' rejected with code 'INVALID_ARGUMENT': Expected to evaluate decision '2251799813685274', but no decision found for key '2251799813685274'
=====================================
Test case 'evaluate decision with decision key' failed, printing container logs up until now
=====================================
Container logs
NOTE: these logs are for all tests in the same suite!
=====================================
=== CONT  TestZbctlWithInsecureGateway/TestCommonCommands
    suite.go:87: test panicked: runtime error: slice bounds out of range [8:6]
        goroutine 108 [running]:
        runtime/debug.Stack()
        	/opt/hostedtoolcache/go/1.19.9/x64/src/runtime/debug/stack.go:24 +0x65
        github.com/stretchr/testify/suite.failOnPanic(0xc0003fe1a0, {0xc97800, 0xc0003e8150})
        	/home/runner/work/zeebe/zeebe/clients/go/vendor/github.com/stretchr/testify/suite/suite.go:87 +0x3b
        github.com/stretchr/testify/suite.Run.func1.1()
        	/home/runner/work/zeebe/zeebe/clients/go/vendor/github.com/stretchr/testify/suite/suite.go:183 +0x252
        panic({0xc97800, 0xc0003e8150})
        	/opt/hostedtoolcache/go/1.19.9/x64/src/runtime/panic.go:884 +0x212
        github.com/camunda/zeebe/clients/go/v8/internal/containersuite.sanitizeDockerLogs({0xc000466000?, 0xc000[41](https://github.com/camunda/zeebe/actions/runs/5188654032/jobs/9352588708?pr=11962#step:7:42)4000?})
        	/home/runner/work/zeebe/zeebe/clients/go/internal/containersuite/containerSuite.go:116 +0x26b
        github.com/camunda/zeebe/clients/go/v8/internal/containersuite.printFailedContainerLogs({0xe2e080, 0xc0005004d0})
        	/home/runner/work/zeebe/zeebe/clients/go/internal/containersuite/containerSuite.go:96 +0x325
        github.com/camunda/zeebe/clients/go/v8/internal/containersuite.(*ContainerSuite).PrintFailedContainerLogs(0xe26b40?)
        	/home/runner/work/zeebe/zeebe/clients/go/internal/containersuite/containerSuite.go:175 +0x3d
        github.com/camunda/zeebe/clients/go/v8/cmd/zbctl.(*integrationTestSuite).TestCommonCommands(0xc000014098)
        	/home/runner/work/zeebe/zeebe/clients/go/cmd/zbctl/main_test.go:284 +0x1eb
        reflect.Value.call({0xc00008b380?, 0xc000014df8?, 0x9f3346?}, {0xcebd69, 0x4}, {0xc000073e70, 0x1, 0xc000073d80?})
        	/opt/hostedtoolcache/go/1.19.9/x64/src/reflect/value.go:584 +0x8c5
        reflect.Value.Call({0xc00008b380?, 0xc000014df8?, 0xc000014098?}, {0xc000073e70?, 0x7f6889b722a8?, 0xd0?})
        	/opt/hostedtoolcache/go/1.19.9/x64/src/reflect/value.go:368 +0xbc
        github.com/stretchr/testify/suite.Run.func1(0xc0003fe1a0)
        	/home/runner/work/zeebe/zeebe/clients/go/vendor/github.com/stretchr/testify/suite/suite.go:197 +0x4b6
        testing.tRunner(0xc0003fe1a0, 0xc000507d40)
        	/opt/hostedtoolcache/go/1.19.9/x64/src/testing/testing.go:1[44](https://github.com/camunda/zeebe/actions/runs/5188654032/jobs/9352588708?pr=11962#step:7:45)6 +0x10b
        created by testing.(*T).Run
        	/opt/hostedtoolcache/go/1.19.9/x64/src/testing/testing.go:1493 +0x35f
2023/06/06 12:49:39 🐳 Terminating container: [46](https://github.com/camunda/zeebe/actions/runs/5188654032/jobs/9352588708?pr=11962#step:7:47)2d1f7ea6c9
2023/06/06 12:[49](https://github.com/camunda/zeebe/actions/runs/5188654032/jobs/9352588708?pr=11962#step:7:50):39 🚫 Container terminated: 462d1f7ea6c9
--- FAIL: TestZbctlWithInsecureGateway (66.[51](https://github.com/camunda/zeebe/actions/runs/5188654032/jobs/9352588708?pr=11962#step:7:52)s)
    --- FAIL: TestZbctlWithInsecureGateway/TestCommonCommands (5.93s)
        --- PASS: TestZbctlWithInsecureGateway/TestCommonCommands/print_help (0.02s)
        --- PASS: TestZbctlWithInsecureGateway/TestCommonCommands/print_version (0.01s)
        --- PASS: TestZbctlWithInsecureGateway/TestCommonCommands/missing_insecure_flag (0.01s)
        --- PASS: TestZbctlWithInsecureGateway/TestCommonCommands/using_insecure_env_var (0.02s)
        --- PASS: TestZbctlWithInsecureGateway/TestCommonCommands/using_json_flag (0.02s)
        --- PASS: TestZbctlWithInsecureGateway/TestCommonCommands/using_json_flag_and_host_and_port_arguments (0.02s)
        --- PASS: TestZbctlWithInsecureGateway/TestCommonCommands/deploy_processes (0.71s)
        --- PASS: TestZbctlWithInsecureGateway/TestCommonCommands/deploy_resources (0.44s)
        --- PASS: TestZbctlWithInsecureGateway/TestCommonCommands/create_instance_with_process_id (1.16s)
        --- PASS: TestZbctlWithInsecureGateway/TestCommonCommands/create_instance_with_process_key (1.09s)
        --- PASS: TestZbctlWithInsecureGateway/TestCommonCommands/evaluate_decision_with_decision_id (1.30s)
        --- FAIL: TestZbctlWithInsecureGateway/TestCommonCommands/evaluate_decision_with_decision_key (1.12s)
FAIL
FAIL	github.com/camunda/zeebe/clients/go/v8/cmd/zbctl	66.[54](https://github.com/camunda/zeebe/actions/runs/5188654032/jobs/9352588708?pr=11962#step:7:55)0s
?   	github.com/camunda/zeebe/clients/go/v8/cmd/zbctl/internal/commands	[no test files]
?   	github.com/camunda/zeebe/clients/go/v8/internal/containersuite	[no test files]
?   	github.com/camunda/zeebe/clients/go/v8/internal/embedded	[no test files]
?   	github.com/camunda/zeebe/clients/go/v8/internal/mock_pb	[no test files]
```

```
Warning:  Flakes: 
Warning:  io.camunda.zeebe.engine.processing.deployment.MultiPartitionDeploymentLifecycleTest.shouldDistributeDmnResources
Error:    Run 1: MultiPartitionDeploymentLifecycleTest.shouldDistributeDmnResources:111 ? StreamWrapper No event found matching the criteria
[INFO]   Run 2: PASS
[INFO] 
Warning:  io.camunda.zeebe.engine.processing.message.MessageCorrelationMultiplePartitionsTest.shouldOpenMessageSubscriptionsOnSamePartitionsAfterRestart
Error:    Run 1: MessageCorrelationMultiplePartitionsTest.init:63 ? StreamWrapper No event found matching the criteria
[INFO]   Run 2: PASS
[INFO] 
Warning:  io.camunda.zeebe.engine.processing.message.MessageStartEventSubscriptionMultiplePartitionsTest.shouldOpenMessageStartEventSubscriptionOnAllPartitions
Error:    Run 1: MessageStartEventSubscriptionMultiplePartitionsTest.shouldOpenMessageStartEventSubscriptionOnAllPartitions:33 ? StreamWrapper No event found matching the criteria
[INFO]   Run 2: PASS
```

File: qa/integration-tests/src/test/java/io/camunda/zeebe/it/util/GrpcClientRule.java [hunk 1]
```diff
@@ -104,12 +105,21 @@ public ZeebeClient getClient() {
...
+                  .withRecordKey(key)
+                  .exists());
+    } else {
+      waitUntil(
+          () ->
+              RecordingExporter.deploymentRecords()
+                  .withIntent(DeploymentIntent.CREATED)
+                  .withRecordKey(key)
+                  .exists());
+    }
```
Comment (korthout) [inline qa/integration-tests/src/test/java/io/camunda/zeebe/it/util/GrpcClientRule.java]: 🔧 We could instead use the RecordingExporter to find the DeploymentIntent.CREATED on all partitions and then check that it exists on all partitions. This is slightly more complex but semantically better fitting. Alternatively, we could check CREATED on 1 partition, and check that the DISTRIBUTION of that key **and** intent was FINISHED

File: protocol-test-util/src/main/java/io/camunda/zeebe/test/broker/protocol/commandapi/PartitionTestClient.java [hunk 2]
```diff
@@ -100,7 +100,7 @@ public long deploy(final BpmnModelInstance process) {
 
     TestUtil.waitUntil(
         () ->
-            RecordingExporter.deploymentRecords(DeploymentIntent.FULLY_DISTRIBUTED)
+            RecordingExporter.deploymentRecords(DeploymentIntent.CREATED)
```
Comment (korthout) [inline protocol-test-util/src/main/java/io/camunda/zeebe/test/broker/protocol/commandapi/PartitionTestClient.java:103]: ❌ This method is used in multi-partition scenarios and the change may lead to flaky tests

`FULLY_DISTRIBUTED` and `CREATED` are not equivalent when this method is used in a multi-partition scenario. That is the case in `ExporterManagerPartitionTest.shouldRunExporterForEveryPartition` and in `UniqueKeyFormatTest.shouldStartProcessInstanceAtNoneStartEvent`.

We should either verify that the distribution of the deployment has FINISHED, or better still: verify that Deployment is CREATED on all partitions.

File: qa/integration-tests/src/test/java/io/camunda/zeebe/it/clustering/DeploymentClusteredTest.java [hunk 3]
```diff
@@ -126,8 +126,7 @@ public void shouldRedistributeDeploymentWhenDeploymentPartitionIsRestarted() {
     final var processDefinitionKey = deploymentEvent.getKey();
 
     assertThat(
-            RecordingExporter.deploymentDistributionRecords()
-                .withIntent(DeploymentDistributionIntent.COMPLETED)
+            RecordingExporter.deploymentRecords(DeploymentIntent.CREATED)
```
Comment (korthout) [inline qa/integration-tests/src/test/java/io/camunda/zeebe/it/clustering/DeploymentClusteredTest.java:129]: ❌ These checks are not equivalent. Let's use `CommandDistribution.ACKNOWLEDGED` (this means the other partition has received it and the distributing partition is aware of this. This scenario should be adjusted for several cases in a83cd854b114bd26f8308fccacf30523d3075b03

**Review by korthout — CHANGES_REQUESTED** [review]
Thanks @remcowesterhoud Really great work ❤️ We're almost there 🏁 

❌ I'll need to make another pass, but these were my findings on the latest commits. Please have a look at my requested changes and suggestions.

File: qa/integration-tests/src/test/java/io/camunda/zeebe/it/util/GrpcClientRule.java (hunk 1 above)
Comment (remcowesterhoud) [reply qa/integration-tests/src/test/java/io/camunda/zeebe/it/util/GrpcClientRule.java]: > We could instead use the RecordingExporter to find the DeploymentIntent.CREATED on all partitions and then check that it exists on all partitions

You mean something like this?

```java
    for (final Integer partition : getPartitions()) {
      waitUntil(
          () ->
              RecordingExporter.deploymentRecords()
                  .withIntent(DeploymentIntent.CREATED)
                  .withPartitionId(partition)
                  .withRecordKey(key)
                  .exists());
    }
```

Comment (korthout) [reply qa/integration-tests/src/test/java/io/camunda/zeebe/it/util/GrpcClientRule.java]: No, that wouldn't really scale well. We can do it one check:

```java
Assertions.assertThat(
  RecordingExporter.deploymentRecords()
    .withIntent(DeploymentIntent.CREATED)
    .withRecordKey(key)
    .limit(getPartitions().size())
).hasPartitionIds(getPartitions()) 
```

File: engine/src/main/java/io/camunda/zeebe/engine/processing/common/CommandDistributionBehavior.java [hunk 4]
```diff
@@ -60,32 +64,34 @@ public <T extends UnifiedRecordValue> void distributeCommand(final TypedRecord<T
...
         CommandDistributionIntent.DISTRIBUTING,
         new CommandDistributionRecord()
             .setPartitionId(partition)
-            .setValueType(command.getValueType())
+            .setValueType(valueType)
             .setIntent(command.getIntent()));
 
     sideEffectWriter.appendSideEffect(
         () -> {
+          System.out.println("LOGJE " + partition);
```
Comment (korthout) [inline engine/src/main/java/io/camunda/zeebe/engine/processing/common/CommandDistributionBehavior.java:94]: 🤡

File: engine/src/test/java/io/camunda/zeebe/engine/util/TestInterPartitionCommandSender.java [hunk 5]
```diff
@@ -55,7 +57,10 @@ public void sendCommand(
       entry = LogAppendEntry.of(metadata, command);
     }
 
-    writer.tryWrite(entry);
+    final Either<WriteFailure, Long> either = writer.tryWrite(entry);
+    if (either.isLeft()) {
+      System.out.println("LOGJE " + either.getLeft());
```
Comment (korthout) [inline engine/src/test/java/io/camunda/zeebe/engine/util/TestInterPartitionCommandSender.java:62]: 🤡

File: qa/integration-tests/src/test/java/io/camunda/zeebe/it/startup/MultiPartitionRecoveryTest.java [hunk 6]
```diff
@@ -85,7 +85,7 @@ public void shouldDistributeDeploymentAfterRestart() {
     reprocessingTrigger.accept(this);
 
     // then
-    clientRule.waitUntilDeploymentIsDone(deploymentEvent.getKey());
+    clientRule.waitUntilDeploymentDistributionIsDone(deploymentEvent.getKey());
```
Comment (korthout) [inline qa/integration-tests/src/test/java/io/camunda/zeebe/it/startup/MultiPartitionRecoveryTest.java:88]: 🔧 Let's revert this name change, it's no longer fitting

**Review by korthout — APPROVED** [review]
🎯 It may have taken some time until we were ready to complete this pull request due to shifting priorities and new findings, but now we're finally there 🏁 

🚀 With this, we have the first command distributed using this mechanism for reliable inter-partition communication, enabling lots of reuse in the future.

👏 Thanks @remcowesterhoud, for all the great efforts. It was a pleasure to work with you on this and to wrap it up together.

LGTM 👍 

bors merge

**Pull request merged**